### ISSUE TITLE
Not able to upload svg file

### ISSUE THREAD

**Issue by rachelwalraven** [joomla/joomla-cms#38532]
After the update to 4.2 it's no longer possible to upload a SVG file.

### Steps to reproduce the issue
Set media options to allow svg files, svg image files and image/svg+xml mimetype
![screen shot 2022-08-19 at 10 12 39](https://issues.joomla.org/uploads/1/545d67c91d028c002441e3b34061ac3c.jpg)
Go to media manager
Upload a svg file

### Expected result
svg file is uploaded and shows in media manager

### Actual result
Error is shown "Unable to upload file"

**Comment by ChristineWk** [comment]
See here please: #38530<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38532">issues.joomla.org/tracker/joomla-cms/38532</a>.</sub>

**Comment by brianteeman** [comment]
@ChristineWk that pull request is just to improve the message. It does NOT resolve the fact that something has changed and all svg uploads are now prevented

**Comment by brianteeman** [comment]
I have found the problem causing this and am working on a pull request

Can someone mark this a release blocker please @roland-d @fancyFranci @richard67

**Comment by ChristineWk** [comment]
Yes, thanks Brian, that's why I didn't close the issue. Just pointed it out for additional information only.

**Comment by brianteeman** [comment]
ok this doesnt make much sense to me. If I understand the code from @bembelimen correctly in the MediaHelper then we are using the enshrined/sanitizer library to try and make the svg is safe and if the library fails to make it safe then we reject it



bad1, bad3, bad4 - these should be blocked from uploading and they currently are
good, good2, good3, good4, good5 - these should be uploaded and they currently are
example1, example2, example3, example4, example5, example6 - these should be uploaded and they are not

cleaned-example6.svg
As a further test I took example1 and example6 and passed them through the online demo of the sanitizer at https://svg.enshrined.co.uk/ and then tried to upload the sanitized result. They still failed.

I hope this information and the zip of images helps someone better than me to resolve this
[svg.zip](https://github.com/joomla/joomla-cms/files/9381802/svg.zip)

**Comment by brianteeman** [comment]
The more I look at this the more confused I am. svgsanitize is designed to _clean_ an upload svg but we appear to be using it as a validator only

**Comment by obuisard** [comment]
Yes, the sanitizer is used to check the files, it is not used to clean them, even though it has that capability. The problem with cleaning files and use the cleaned version is that it my not have the intended results. For instance, the use of namespaces in the file. Removed, it could break the SVG output.

**Comment by obuisard** [comment]
The media manager can be used in the frontend so allowing SVG files that may have concerned issues should not be uploaded. However, if an administrator of the site knowingly puts a SVG on his/her site through FTP, for instance, nothing prevents him/her to. The admin is responsible for making sure the file is safe.

**Comment by obuisard** [comment]
Now we need to educate the user so he knows why the file is harmful. Hence better wording on the messages we send and proper documentation.

**Comment by brianteeman** [comment]
No not at all. Use svgsanitize as designed.

**Comment by brianteeman** [comment]
Telling a user that an svg file from a reputable source/application is harmful when it is not doesn't help anyone at all.

**Comment by brianteeman** [comment]
further tests with the svg files I supplied above and using the svg-scanner that is part of the package suggest that the scanner is crap. try it yourself and see how it will fail almost any svg that is not minified

**Comment by N6REJ** [comment]
> Telling a user that an svg file from a reputable source/application is harmful when it is not doesn't help anyone at all.

100% spot on.  Here are 2 images created via inkscape which is the FOSS eqv to Adobe illustrator and both will fail being uploaded.
![Bearsampp-header-logo](https://user-images.githubusercontent.com/1850089/185713330-ab14efe5-98ce-4510-8a25-164346912072.svg)
![Bearsampp-logo-128x128](https://user-images.githubusercontent.com/1850089/185713333-5b9e09ff-92cb-4d81-ba36-6a6ba335fb0d.svg)
I could supply several more.
The point being is I EXPECT these files to work.  They were created with a well known program designed for this kind of work.

**Comment by brianteeman** [comment]
@N6REJ if you minify them then they will probably work

**Comment by dgrammatiko** [comment]
SVG security is HIGHLY dependant on the way the SVG is used. As an img src the sanitisation is done BY the browser, even unsanitised svgs are fine. Read the specs: https://svgwg.org/svg2-draft/conform.html#examples

**Comment by SniperSister** [comment]
> As an img src the sanitisation is done BY the browser, even unsanitised svgs are fine

The thing is: can we safely assume that an uploaded SVG is always used in the img-src context? I don't think so.

**Comment by brianteeman** [comment]
No we can't but we can at least use the sanitize library as a sanitizer. Its clearly not great as a valaidator

**Comment by dgrammatiko** [comment]
> we safely assume that an uploaded SVG is always used in the img-src context

No, the CMS doesn't have this flexibility and neither I posted the reference to imply this. I just wanted to point out that if SVGs are used as images (with src attribute pointing to the url of the image) there's no need for sanitisation. It's unrealistic to assume that this would be the only way people will use them as they can have a field and then embed the svg in the html, or in the css or...

> No we can't but we can at least use the sanitize library as a sanitizer

Maybe it's easier to do the sanitisation on the client before even it reaches the server. The code from the [SVGOM](https://jakearchibald.github.io/svgomg/) could easily imported and executed when the mime type === svg/... https://github.com/jakearchibald/svgomg/tree/main/src/js

**Comment by brianteeman** [comment]
> Maybe it's easier to do the sanitisation on the client before even it reaches the server. 

That's still telling someone that the svg from their reputable source is vulnerable just because we are using a library incorrectly

**Comment by dgrammatiko** [comment]
My point was that IF you can guarantee that svgs will only be used as image tags then there's no need for any sanitisation, proof is that the images from the svg.zip that @brianteeman posted above were uploaded without any changes on GitHub because the only way GH will ever use them is as img tags. In short this might give you another perspective...

![bad1](https://user-images.githubusercontent.com/3889375/185746780-7bc78e1c-fabe-4ad6-9905-6085b1aaa84b.svg)
![bad3](https://user-images.githubusercontent.com/3889375/185746781-b90edfeb-6c10-41ab-972c-9a1b22979295.svg)
![bad4](https://user-images.githubusercontent.com/3889375/185746782-f8bb0256-b383-41b5-9c42-e86919bb34e9.svg)
![cleaned-example1](https://user-images.githubusercontent.com/3889375/185746784-669dae0e-75cd-438a-bf86-9ece621aa000.svg)
![cleaned-example6](https://user-images.githubusercontent.com/3889375/185746786-8052c460-03a0-4ac9-95b4-f2e07e257322.svg)
![example1](https://user-images.githubusercontent.com/3889375/185746788-7c1c39c4-2680-448b-931b-f6011fe320b8.svg)
![example2](https://user-images.githubusercontent.com/3889375/185746789-c2672bcb-263b-466a-8331-000d6df8e33d.svg)
![example3](https://user-images.githubusercontent.com/3889375/185746790-f9944b31-c89f-4728-97dd-b7a36cdc5c07.svg)
![example4](https://user-images.githubusercontent.com/3889375/185746791-9ec88d77-279c-4c96-874c-f6a5c1de3716.svg)
![example-6](https://user-images.githubusercontent.com/3889375/185746792-efd30bd2-3e92-41ad-9a78-24b84d5638a1.svg)
![example5](https://user-images.githubusercontent.com/3889375/185746795-3a76ea33-d9be-4a6f-9848-c64537ecc51a.svg)
![good2](https://user-images.githubusercontent.com/3889375/185746796-40a55f75-7673-448f-8fce-ec521f688c94.svg)
![good](https://user-images.githubusercontent.com/3889375/185746797-9c05090e-f384-44ba-a7f5-d5e9c08bdfda.svg)
![good3](https://user-images.githubusercontent.com/3889375/185746798-0aa1c84b-3b67-4150-bb9b-95c3d2a95745.svg)
![good4](https://user-images.githubusercontent.com/3889375/185746799-5471b5fd-f6b8-4dfd-b64b-dffff4d5198f.svg)
![good5](https://user-images.githubusercontent.com/3889375/185746800-6837fbb6-a959-4742-82c7-62075202050f.svg)

**Comment by obuisard** [comment]
> SVG security is HIGHLY dependant on the way the SVG is used. As an img src the sanitisation is done BY the browser, even unsanitised svgs are fine. Read the specs: https://svgwg.org/svg2-draft/conform.html#examples

My understanding is that the img tag disables scripting (SVG files are 'run' in a browser sandbox), but the files are not sanitized. Therefore, anyone downloading the file from your site get the original file. 
Is that right @dgrammatiko?

**Comment by brianteeman** [comment]
The security issue occurs if a malicious svg could be uploaded (which is a low privilege task) it could then be accessed directly www.example.com/images/filename.svg . Image files are not prevented from being loaded directly

**Comment by brianteeman** [comment]
@roland-d please explain your justification for removing the release blocker label.

**Comment by N6REJ** [comment]
@brianteeman yes that did work.

**Comment by brianteeman** [comment]
@N6REJ its because one of the rules in svgsanitize is to mark a space as an error. We could easily allow spaces just as we allow comments.

**Comment by N6REJ** [comment]
looks like if we'd just leave it alone and let its clean it all would be well

**Comment by brianteeman** [comment]
yes. but thats not we're doing

**Comment by bembelimen** [comment]
> @N6REJ its because one of the rules in svgsanitize is to mark a space as an error. We could easily allow spaces just as we allow comments although I still prefer to use the library as it was intended

Not sure what you mean with "spaces" but spaces are not marked as invalid. An attribute (xml:space) is marked as invalid due to a bug: https://github.com/darylldoyle/svg-sanitizer/issues/64

**Comment by brianteeman** [comment]
```
 
PS C:\laragon\www\j4\libraries\vendor\enshrined\svg-sanitize\src> php .\svg-scanner.php .\svg\example2.svg
{
    "totals": {
        "errors": 3
    },
    "files": {
        ".\\svg\\example2.svg": {
            "errors": 3,
            "messages": [
                {
                    "message": "Suspicious attribute 'space'",
                    "line": 3
                },
                {
                    "message": "Suspicious attribute 'space'",
                    "line": 8
                },
                {
                    "message": "Suspicious node 'svg'",
                    "line": -1
                }
            ]
        }
    }
}
```

ah I see what you mean now about xml:space and that it was just coincidence on the files it was reported. Perhaps until the library is updated you could whitelist it in the same way as you did the comments?

**Comment by brianteeman** [comment]
Found a few more bugs upstream. Please test #38545 as this removes a few more false positives that are preventing upload

**Comment by N6REJ** [comment]
> > @N6REJ its because one of the rules in svgsanitize is to mark a space as an error. We could easily allow spaces just as we allow comments although I still prefer to use the library as it was intended
> 
> Not sure what you mean with "spaces" but spaces are not marked as invalid. An attribute (xml:space) is marked as invalid due to a bug: [darylldoyle/svg-sanitizer#64](https://github.com/darylldoyle/svg-sanitizer/issues/64)

@brianteeman said that not me.

**Comment by dgrammatiko** [comment]
> My understanding is that the img tag disables scripting (SVG files are 'run' in a browser sandbox), but the files are not sanitized. Therefore, anyone downloading the file from your site gets the original file. Is that right @dgrammatiko?

@obuisard so, the SVG is never a threat or in anyway problematic for the server side, it's just a static file! For the client side ONLY if the SVG is embedded into the page there are potential problems. The images folder (or any folder that the media manager is managing) is supposed to be only for static assets (jpg, png, webp, avif, mp4, mov, pdf, etc) and the only way that these files are used for the CMS is ALWAYS as image, video, audio or object. There is no way that an SVG will ever be embedded in the page UNLESS a user or a developer does specifically that. In short the whole fear for SVGs is not justified. On top of that if SVGs are problematic because they might introduce XSS vulnerabilities then there are already documented such vulnerabilities in other parts of the software that got a status `won't fix` or `not an issue`. It seems to me that there are 2 rules...

**Comment by roland-d** [comment]
> @roland-d please explain your justification for removing the release blocker label.

There are several reasons I did it at the time:
- This issue is not blocking the 4.2.1 release
- There was no PR with a fix at the time
- SVGs can still be uploaded although I agree not all of them

Now following the discussion here, can't we make it an option that users can decide for themselves to allow SVG uploads? By default we do what we do now but if users want to open it up they can. They can upload SVGs Joomla considers bad anyway via FTP/Filemanager/SSH.

**Comment by brianteeman** [comment]
not having a pr is not a reason to remove a release blocker 🤦 

as release lead I respectfully suggest that you do follow the discussion or there is no way yu can actually make a good faith decision.

**Comment by roland-d** [comment]
> not having a pr is not a reason to remove a release blocker :facepalm:

Fair enough, I will put the release blocker back

> as release lead I respectfully suggest that you do follow the discussion or there is no way yu can actually make a good faith decision.

I am following the discussion, both here and on Glip

**Comment by crystalenka** [comment]
I actually ran into this issue yesterday trying to test the PR for media manager thumbnails. 

At a bare minimum, the error text should be expanded. I thought Joomla was broken. Explaining that the SVG could be potentially harmful, and maybe linking to a place to minify it, would be extremely helpful.

Even better would be to offer the user the option to sanitize the SVG with a note it may behave in an unexpected way. That way it's an opt-in action, and if it breaks the user knows why.

**Comment by brianteeman** [comment]
As https://github.com/joomla/joomla-cms/pull/38545 has now been merged there should be a big drop in failed svg

**Comment by N6REJ** [comment]
> Even better would be to offer the user the option to sanitize the SVG with a note it may behave in an unexpected way. That way it's an opt-in action, and if it breaks the user knows why.

100% agree with that

**Comment by HDInfautre** [comment]
Hello

I would point out that I can send svg files via FTP, then use them in the media manager (joomla 4.2.5).
Most of the time, it is impossible to download them from the media manager. The error message does not give much information.
Is this fixed for you?
Is this something that can be improved in a future version of Joomla?

Regards<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38532">issues.joomla.org/tracker/joomla-cms/38532</a>.</sub>

**Comment by fancyFranci** [comment]
This issue is not blocking any release, so I'm removing the release blocker label. Meanwhile a helpful fix was merged and workarounds via FTP exist. Of course I would still welcome an improvement like mentionend above.

**Comment by dgrammatiko** [comment]
@fancyFranci is this still valid? When people tested https://github.com/joomla/joomla-cms/pull/39586 nobody mentioned that they were unable to upload SVG files

**Comment by obuisard** [comment]
Don't forget, it's fixed in 4.3, not 4.2

**Comment by fancyFranci** [comment]
Ok nice, I missed that PR. Then the issue could be closed indeed.

**Comment by brianteeman** [comment]
@obuisard what PR are you referring to?

My pr #38545 addressed some issues but that was for 4.2. Testing 4.3 with the same svg I have provided here already still shows  multiple svg being rejected

There is still the outstanding problem as mentioned by @crystalenka 

> I actually ran into this issue yesterday trying to test the PR for media manager thumbnails.
> 
> At a bare minimum, the error text should be expanded. I thought Joomla was broken. Explaining that the SVG could be potentially harmful, and maybe linking to a place to minify it, would be extremely helpful.
> 
> Even better would be to offer the user the option to sanitize the SVG with a note it may behave in an unexpected way. That way it's an opt-in action, and if it breaks the user knows why.

The error text is still the unhelpful 
> Error
> Unable to upload file.

**Comment by brianteeman** [comment]
> @fancyFranci is this still valid? When people tested #39586 nobody mentioned that they were unable to upload SVG files

yes they did (https://github.com/joomla/joomla-cms/pull/39586#issuecomment-1383453306)

**Comment by dgrammatiko** [comment]
> yes they did (https://github.com/joomla/joomla-cms/pull/39586#issuecomment-1383453306)

Actually that was a problem with PHP7.x reporting the mime type differently and solved with the commit https://github.com/joomla/joomla-cms/pull/39586/commits/dfcf2056a6c26923b851ebf3f533ede63834c058

My point **wasn't** that my PR fixed the problem (yours did, I didn't;t claimed that) but asking if this is still valid. Also for the error reporting @Fedik has a PR https://github.com/joomla/joomla-cms/pull/38536...

**Comment by obuisard** [comment]
Brian @brianteeman I was thinking of the fix from Dimitris, as he explained in the previous comment.
However, it does not prevent SVG files from being denied drop into the Media Manager if the file is considered 'suspicious' by the sanitizer script we are using. 
You have eased the rules (PR #38545) and the messaging is more user friendly (PR #38536), mime types are properly handled. 
However, we will never totally 'fix' the SVG uploads because we don't use the sanitizer to actually 'sanitize' files.

**Comment by brianteeman** [comment]
@obuisard except that the messaging improvement has not been merged yet

> However, we will never totally 'fix' the SVG uploads because we don't use the sanitizer to actually 'sanitize' files.

any why don't we?

**Comment by Orgoth** [comment]
Joomla 5.2.3 it is impossible to upload an SVG!
I only get a useless Message:

> Unable to upload file.

![Image](https://github.com/user-attachments/assets/3d61a633-9b26-47f7-b3cf-025d36481913)

Extensions and mime-type is whitelisted within the media-manager config.
The Admin should be able to allow SVG if there are no possible upload functions for guests and other users.

**Comment by Orgoth** [comment]
Sanitizer is breaking the upload for legit svg's.

https://github.com/darylldoyle/svg-sanitizer/issues/111

I have modified the Media-Manager to report the errors instead of completely hiding them!

```php
Factory::getApplication()->enqueueMessage(Text::_('JLIB_MEDIA_ERROR_WARNIEXSS').'-'.json_encode($svgErrors), 'error');
```

```
Possible IE XSS Attack found.
[{
    "message": "Suspicious attribute 'dominant-baseline'",
    "line": 87
}]
```

**Comment by Orgoth** [comment]
@brianteeman please update svg-sanitizer.

https://github.com/darylldoyle/svg-sanitizer/pull/112

**Comment by richard67** [comment]
> [@brianteeman](https://github.com/brianteeman) please update svg-sanitizer in joomla.
> 
> [darylldoyle/svg-sanitizer#112](https://github.com/darylldoyle/svg-sanitizer/pull/112)

New version is https://github.com/darylldoyle/svg-sanitizer/releases/tag/0.21.0

@brianteeman As you were pinged: Do you want to make a PR to update that dependency? If not, I can do on weekend. I think it's a bug fix and should to into 5.2-dev.

**Comment by brianteeman** [comment]
I've lost the will to do anything with svg - someone else will have to make the pr

**Comment by richard67** [comment]
> [@brianteeman](https://github.com/brianteeman) please update svg-sanitizer in joomla.
> 
> [darylldoyle/svg-sanitizer#112](https://github.com/darylldoyle/svg-sanitizer/pull/112)

@Orgoth Please test #44746 . Thanks in advance.

**Comment by Orgoth** [comment]
> > [@brianteeman](https://github.com/brianteeman) please update svg-sanitizer in joomla.
> > [darylldoyle/svg-sanitizer#112](https://github.com/darylldoyle/svg-sanitizer/pull/112)
> 
> [@Orgoth](https://github.com/Orgoth) Please test [#44746](https://github.com/joomla/joomla-cms/pull/44746) . Thanks in advance.

Is my testing still needed, since it is already merged?

**Comment by richard67** [comment]
> Is my testing still needed, since it is already merged?

@Orgoth Not really needed, but if you can confirm that it’s what you suggested and that it works would be nice.

**Comment by Orgoth** [comment]
> > Is my testing still needed, since it is already merged?
> 
> [@Orgoth](https://github.com/Orgoth) Not really needed, but if you can confirm that it’s what you suggested and that it works would be nice.

Tested, the upload works now.  👍

> > Telling a user that an svg file from a reputable source/application is harmful when it is not doesn't help anyone at all.
> 
> 100% spot on. Here are 2 images created via inkscape which is the FOSS eqv to Adobe illustrator and both will fail being uploaded.  I could supply several more. The point being is I EXPECT these files to work. They were created with a well known program designed for this kind of work.

This is, because of Suspicious attribute 'serif:id' on line 1.

```xml
<g id="Page-1" serif:id="Page 1">
```

 ![Bearsampp-header-logo](https://user-images.githubusercontent.com/1850089/185713330-ab14efe5-98ce-4510-8a25-164346912072.svg)

Here I have asked whether it is also possible to consider attributes with namespace as a whitelist.

https://github.com/darylldoyle/svg-sanitizer/blob/master/src/data/AllowedAttributes.php#L312
https://github.com/darylldoyle/svg-sanitizer/issues/113

The attribute "id" is already whitelisted.

The other image is possible to upload with the latest fix.
https://github.com/joomla/joomla-cms/issues/38532#issuecomment-2599729238
